oeps/oep-0012.rst

```diff
@@ -0,0 +1,520 @@
+chooses.
+
+Prototype: Instructor Dashboard
+-------------------------------
+
+The final piece of the prototype demonstrates a dynamically generated
+instructor dashboard built purely using component views. The dashboard home
+page makes a request to find all installed instructor component and then
+displays them as a set of cards organized by category:
+
+.. image:: oep-0012/proposed-instructor-dashboard.png
+  :alt: A screen shot from the instructor dashboard prototype, showing
+  how each feature that can be controlled is shown as a card that can
+  be drilled down into to be seen full screen.
+
+Each card has a "View" link that lets the user drill down and see the
+component for the feature being represented. This component is shown full
+screen but with a back button to allow the user to navigate back to the
+dashboard home.
+
+Rationale
+=========
+
+Why are we not using an industry framework?
+-------------------------------------------
+
+It seems that the requirements for pluggable user interfaces are not unique
+to Open edX, and hence other solutions are likely to exist. Our research so
+far has not found a solution that covers the full set of use cases, but that
+might change in the future.
+
+One approach that is similar is Pinterest's solution that can partially render a
+page, and then asynchronously request the missing components and fill them in as
+they become available. They use Mozilla's Nunjucks templating library in Django
+and in particular their `Nunjucks Asynchronous Support`_. However, it appears
+that this only supports the ability to asynchronously return HTML, and doesn't
+allow for the ability to dynamically add CSS and JavaScript. See their blog
+post `How we switched our template rendering engine to React`_ for more
+details.
+
+Another similar use case is HTML5 ads, but those are all implemented as
+iframes. Using iframes is possible for some pluggable situations (such as LTI),
+but it would provide a poor user experience if a complex page was constructed
+from tens or even hundreds of iframes.
+
+Could we use XBlocks instead of introducing Django Component Views?
+-------------------------------------------------------------------
+
+Note that all of the requirements of pluggable user interfaces are supported by
+XBlocks. However, there are several reasons why XBlocks are not always
+appropriate:
+
+* XBlocks need to have a unique usage id so that instances can be referenced.
+
+* XBlocks need to have storage scoped for different purposes (per learner, per
+  course etc).
+
+* XBlocks in the LMS are tied to a course, which doesn't make them great for
+  general user interfaces (e.g. UIs that are global, organization-wide or
+  program-wide).
+
+* XBlocks are persisted while component views are ephemeral. This also means
+  that converting an existing view into an XBlock requires a data migration
+  so that each course is updated to contain a block instance.
```

Comment:
I don't quite understand this comment. Both are rendered on-the-fly. Both sometimes make use of permanent storage. I'm not quite sure I catch the distinction.